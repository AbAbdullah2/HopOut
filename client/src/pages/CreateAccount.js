import React, { useState } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { solid } from '@fortawesome/fontawesome-svg-core/import.macro'

function CreateAccount(props) {
    const [email, setEmail] = useState("");
    const [password, setPassword] = useState("");
    const [confirmPassword, setConfirmPassword] = useState("");
    const [remember, setRemember] = useState(false);
    const [success, setSuccess] = useState(false);
    const [toast, setToast] = useState(false);

    const create = (e) => {
        console.log("create:", e);
        if (email.substring(email.length - 8) !== "@jhu.edu" || password !== confirmPassword || password.length < 6) {
            setSuccess(false); 
            return;
        }
        else setSuccess(true);
        setToast(true);
    }

    return (
        <>
        <div className="flex min-h-full items-center justify-center py-12 px-4 sm:px-6 lg:px-8">
            <div className="w-full max-w-md space-y-8">
            <div>
                <h2 className="mt-6 text-center text-3xl font-bold tracking-tight text-gray-900">
                Create a new HopOut account
                </h2>
            </div>
            <form className="mt-8 space-y-6" action="#" method="POST">
                <input type="hidden" name="remember" defaultValue="true" />
                <div className="-space-y-px rounded-md shadow-sm">
                <div className="mb-3">
                    <label htmlFor="email-address" className="sr-only">
                    Email address
                    </label>
                    <input
                    id="email-address"
                    name="email"
                    type="email"
                    autoComplete="email"
                    required
                    className="relative block w-full appearance-none rounded border border-gray-300 px-3 py-2 text-gray-900 placeholder-gray-500 focus:z-10 focus:border-indigo-500 focus:outline-none focus:ring-indigo-500 sm:text-sm"
                    placeholder="Email address (must end in @jhu.edu)"
                    value={email}
                    onChange={(e) => setEmail(e.target.value)}
                    />
                </div>
                <div>
                    <label htmlFor="password" className="sr-only">
                    Password
                    </label>
                    <input
                    id="password"
                    name="password"
                    type="password"
                    autoComplete="current-password"
                    required
                    className={password.length != 0 && password.length < 6 ? 
                    "relative block w-full appearance-none rounded border border-gray-300 px-3 py-2 text-gray-900 placeholder-gray-500 focus:z-10 focus:border-indigo-500 focus:outline-none focus:ring-indigo-500 sm:text-sm border-red-500" :
                    "relative block w-full appearance-none rounded border border-gray-300 px-3 py-2 text-gray-900 placeholder-gray-500 focus:z-10 focus:border-indigo-500 focus:outline-none focus:ring-indigo-500 sm:text-sm"}
                    placeholder="Password"
                    value={password}
                    onChange={(e) => setPassword(e.target.value)}
                    />
                    {password.length != 0 && password.length < 6 ? <p class="mb-0 text-red-500 text-xs italic">Password must at least 6 characters.</p> : <></>}

                </div>
                <div>
                    <label htmlFor="password" className="sr-only">
                    Confirm password
                    </label>
                    <input
                    id="password"
                    name="password"
                    type="password"
                    autoComplete="current-password"
                    required
                    className="mt-3 relative block w-full appearance-none rounded border border-gray-300 px-3 py-2 text-gray-900 placeholder-gray-500 focus:z-10 focus:border-indigo-500 focus:outline-none focus:ring-indigo-500 sm:text-sm"
                    placeholder="Confirm password"
                    value={confirmPassword}
                    onChange={(e) => setConfirmPassword(e.target.value)}
                    />
                </div>
                </div>

                <div className="flex items-center justify-between">
                <div className="flex items-center">
                    <input
                    id="remember-me"
                    name="remember-me"
                    type="checkbox"
                    className="h-4 w-4 rounded border-gray-300 text-indigo-600 focus:ring-indigo-500"
                    value={remember}
                    onChange={() => setRemember(!remember)}
                    />
                    <label htmlFor="remember-me" className="ml-2 block text-sm text-gray-900">
                    Remember me
                    </label>
                </div>
                </div>

                <div>
                <button
                    type="submit"
                    className="group relative flex w-full justify-center rounded-md border border-transparent bg-indigo-600 py-2 px-4 text-sm font-medium text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2"
                    onClick={(e) => create(e)}
                >
                    <span className="absolute inset-y-0 left-0 flex items-center pl-3">
                    <FontAwesomeIcon icon={solid('lock')} className="h-5 w-5 text-indigo-500 group-hover:text-indigo-400" aria-hidden="true" />
                    </span>
                    Create account
                </button>
                </div>
            </form>
            {toast ?  (
            <div id="toast-default" className="flex justify-center items-center w-full max-w-md p-4 text-gray-500 bg-white rounded-lg shadow dark:text-gray-400 dark:bg-gray-600" role="alert">
                <div className="ml-3 text-sm font-normal">{success ? "Account created successfuly!" : "Something went wrong. Please try again. Can only create account with email ending in \".jhu.edu\"."}</div>
                <button type="button" className="ml-auto -mx-1.5 -my-1.5 bg-white text-gray-400 hover:text-gray-900 rounded-lg focus:ring-2 focus:ring-gray-300 p-1.5 hover:bg-gray-100 inline-flex h-8 w-8 dark:text-gray-600 dark:hover:text-white dark:bg-gray-500 dark:hover:bg-gray-700" data-dismiss-target="#toast-default" aria-label="Close" onClick={()=> setToast(false)}>
                    <span className="sr-only">Close</span>
                    <svg aria-hidden="true" className="w-5 h-5" fill="currentColor" viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg"><path fill-rule="evenodd" d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z" clip-rule="evenodd"></path></svg>
                </button>
            </div>            
            ): <></>}
            </div>
        </div>
        </>
    );
}

export default CreateAccount;