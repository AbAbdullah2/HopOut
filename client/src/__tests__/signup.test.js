import '@testing-library/jest-dom'
import {render, screen, fireEvent} from '@testing-library/react'
import renderer from 'react-test-renderer';
import CreateAccount from '../pages/CreateAccount'

const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
   ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUsedNavigate,
}));
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

test('renders login fields', () => {
  render(<CreateAccount />);

  const emailInput = screen.getByLabelText('Email');
  const nameInput = screen.getByLabelText('Name');
  const passwordInput = screen.getByLabelText('Password');
  const confirmPasswordInput = screen.getByLabelText('Confirm password');
  const signupButton = screen.getByRole('button', { name: 'Create account' });

  expect(emailInput).toBeInTheDocument();
  expect(nameInput).toBeInTheDocument();
  expect(passwordInput).toBeInTheDocument();
  expect(confirmPasswordInput).toBeInTheDocument();
  expect(signupButton).toBeInTheDocument();
});

test('password alerts', () => {    
  render(<CreateAccount />);

  const passwordInput = screen.getByLabelText('Password');
  const confirmPasswordInput = screen.getByLabelText('Confirm password');
  const signupButton = screen.getByRole('button', { name: 'Create account' });

  fireEvent.change(passwordInput, { target: { value: '123' } });
  expect(screen.getByText('Password must be at least 6 characters.')).toBeInTheDocument();
  expect(screen.getByText('Passwords must match.')).toBeInTheDocument();

  fireEvent.change(confirmPasswordInput, { target: { value: '123' } });
  fireEvent.click(signupButton);
});


test('matches snapshot', () => {
  const tree = renderer.create(<CreateAccount/>).toJSON();
  expect(tree).toMatchSnapshot();
})