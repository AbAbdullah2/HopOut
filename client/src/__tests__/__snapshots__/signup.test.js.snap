// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="bg-stone-100 min-h-screen"
>
  <div
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "bottom": 16,
        "left": 16,
        "pointerEvents": "none",
        "position": "fixed",
        "right": 16,
        "top": 16,
        "zIndex": 9999,
      }
    }
  >
    <div
      className="go4109123758"
      style={
        Object {
          "display": "flex",
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "transform": "translateY(0px)",
          "transition": undefined,
        }
      }
    >
      <div
        className="go2072408551"
        style={
          Object {
            "opacity": 0,
          }
        }
      >
        <div
          className="go685806154"
        >
          <div
            className="go1858758034"
          />
          <div
            className="go1579819456"
          >
            <div
              className="go2534082608"
            />
          </div>
        </div>
        <div
          aria-live="polite"
          className="go3958317564"
          role="status"
        >
          Email must end in @jhu.edu or @jh.edu
        </div>
      </div>
    </div>
  </div>
  <div
    className="mx-auto flex flex-col h-full"
  >
    <div
      className="w-full bg-slate-800 p-5 text-white flex flex-row flex-nowrap justify-center text-center items-center"
    >
      <a
        className="mx-2"
        href="/"
      >
        <span
          className="text-4xl font-extrabold italic text-blue-400 mr-1"
        >
          H
        </span>
        <span
          className="text-4xl"
        >
          opOut
        </span>
      </a>
    </div>
    <div
      className="flex items-center justify-center py-6"
    >
      <div
        className="w-full max-w-md"
      >
        <h2
          className="mt-6 text-center text-3xl font-bold tracking-tight text-gray-900"
        >
          Sign Up
        </h2>
        <form
          className="mt-8 space-y-6"
          onSubmit={[Function]}
        >
          <div
            className="rounded-md shadow-sm space-y-3"
          >
            <div>
              <label
                className="sr-only"
                htmlFor="email"
              >
                Email
              </label>
              <input
                autoComplete="email"
                className="relative block w-full appearance-none rounded border border-gray-300 px-3 py-2 text-gray-900 placeholder-gray-500 focus:z-10 focus:border-blue-500 focus:outline-none focus:ring-indigo-500"
                id="email"
                name="email"
                onChange={[Function]}
                placeholder="JHU Email"
                required={true}
                type="email"
                value=""
              />
            </div>
            <div>
              <label
                className="sr-only"
                htmlFor="name"
              >
                Name
              </label>
              <input
                autoComplete="name"
                className="relative block w-full appearance-none rounded border border-gray-300 px-3 py-2 text-gray-900 placeholder-gray-500 focus:z-10 focus:border-blue-500 focus:outline-none focus:ring-blue-500"
                id="name"
                name="name"
                onChange={[Function]}
                placeholder="Name"
                required={true}
                type="name"
                value=""
              />
            </div>
            <div>
              <label
                className="sr-only"
                htmlFor="password"
              >
                Password
              </label>
              <input
                autoComplete="current-password"
                className="relative block w-full appearance-none rounded border border-gray-300 px-3 py-2 text-gray-900 placeholder-gray-500 focus:z-10 focus:border-blue-500 focus:outline-none focus:ring-indigo-500"
                id="password"
                name="password"
                onChange={[Function]}
                placeholder="Password"
                required={true}
                type="password"
                value=""
              />
            </div>
            <div>
              <label
                className="sr-only"
                htmlFor="passwordconfirm"
              >
                Confirm password
              </label>
              <input
                autoComplete="current-password"
                className="relative block w-full appearance-none rounded border border-gray-300 px-3 py-2 text-gray-900 placeholder-gray-500 focus:z-10 focus:border-blue-500 focus:outline-none focus:ring-indigo-500"
                id="passwordconfirm"
                name="passwordconfirm"
                onChange={[Function]}
                placeholder="Confirm password"
                required={true}
                type="password"
                value=""
              />
            </div>
          </div>
          <div>
            <button
              className="group relative flex w-full justify-center rounded-md border border-transparent bg-blue-400 py-2 px-4 text-sm font-medium text-white hover:bg-slate-800 focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2"
              type="submit"
            >
              Create Account
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
